Blockly Factory: Enhanced Block Library UI

quachtina96 opened the pull request:
# Overview
- The Block Library is now a dropdown with the Blockly Factory styling.
- The 'Create New Block' button is now housed in the Block Library dropdown.
- Based on the state of the user in Block Factory, the Save/Delete buttons for the Block Library update content and color to guide the user's actions.
- green buttons indicate user can save or update a saved block
- the red save button indicates that user cannot save the block  (but encourages user to click button to learn why)

## When user first visits Blockly Factory
<img width="1438" alt="screen shot 2016-08-25 at 8 08 41 pm" src="https://cloud.githubusercontent.com/assets/10423718/17992963/92c705d4-6b00-11e6-8d42-2b34d423b506.png">
### Warning text is set on the block.
<img width="758" alt="screen shot 2016-08-25 at 7 52 53 pm" src="https://cloud.githubusercontent.com/assets/10423718/17992969/a1b4498a-6b00-11e6-93b1-f0b8fa1a9ac1.png">
### Those that don't notice should click the 'Save' Button for a message to guide them on how to use the tool.
<img width="1036" alt="screen shot 2016-08-25 at 7 52 43 pm" src="https://cloud.githubusercontent.com/assets/10423718/17992970/a1b669f4-6b00-11e6-9eeb-22264d9d97a1.png">

## After changing the name of the block, buttons are updated and the button is green!
<img width="1438" alt="screen shot 2016-08-25 at 8 18 53 pm" src="https://cloud.githubusercontent.com/assets/10423718/17993012/2d055b3c-6b01-11e6-93f0-a079a14c3bcd.png">

## User might edit the block
The button turns green to indicate that they should save.
<img width="1429" alt="screen shot 2016-08-25 at 8 22 29 pm" src="https://cloud.githubusercontent.com/assets/10423718/17993085/b03f2d34-6b01-11e6-8a8f-8b641e60aabd.png">

## When the user saves edits, the update button changes back to grey and is disabled.
<img width="1440" alt="screen shot 2016-08-25 at 8 16 00 pm" src="https://cloud.githubusercontent.com/assets/10423718/17993006/1ff7cdd0-6b01-11e6-959a-9c446c68d04d.png">

## If the user deletes the block, they now have a green save button!
<img width="1440" alt="screen shot 2016-08-25 at 8 16 07 pm" src="https://cloud.githubusercontent.com/assets/10423718/17993007/1ff88d4c-6b01-11e6-84f4-6dff28672be1.png">

## Checking the library indicates that the 'color_blue' block is saved and currently selected.
<img width="1437" alt="screen shot 2016-08-25 at 8 16 51 pm" src="https://cloud.githubusercontent.com/assets/10423718/17993003/1ff753be-6b01-11e6-9ff5-6ef2b0112339.png">






consistent button name on default

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly/603)
<!-- Reviewable:end -->

quachtina96 commented:
@vicng @picklesrus @evd2014

picklesrus commented:
Didn't quite get through this, but losing internet for a bit shortly so wanted to send comments so far.  Definitely like the new UI!

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[demos/blocklyfactory/app_controller.js, line 296 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7ek2M5-SOOcITbbET:-KQ7ek2M5-SOOcITbbEU:by3zr6q) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/app_controller.js#L296)):*
> ```JavaScript
>         // stay on Block Factory Tab.
>         this.setSelected_(AppController.BLOCK_FACTORY);
>         this.lastSelectedTab == AppController.BLOCK_FACTORY;
> ```

Do you mean = here?

---

*[demos/blocklyfactory/block_library_controller.js, line 55 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7h3GfU95vQhbRBeaP:-KQ7h3GgZzU6H4KTRxYx:bz09fyf) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L55)):*
> ```JavaScript
>   this.storage = opt_blockLibraryStorage || new BlockLibraryStorage(this.name);
>   // Id of the div that holds the block library view.
>   this.blockLibraryViewDivID = 'dropdownDiv_blockLib';
> ```

I'm wondering if you need this. Mostly because I think the places where you use it (e.g. hiding/showing the list), you probably should just have a view method for hide/show.  

---

*[demos/blocklyfactory/block_library_controller.js, line 190 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7hbh7CSJnxPJcHS4S:-KQ7hbh86B1_f_AuO4Ub:b-r6fwbi) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L190)):*
> ```JavaScript
>     this.view.addOption(blockType, false);
>   }
>   this.addOptionSelectHandlers();
> ```

should this be done by the view? (caveat: I haven't looked at what addoptionselectHandlers() does yet)

---

*[demos/blocklyfactory/block_library_controller.js, line 285 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7g4UvW_TxjYrbqG2U:-KQ7g4UvW_TxjYrbqG2V:bnbbke3) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L285)):*
> ```JavaScript
>       // warn user that proceeding to create a new block will cause them to lose
>       // their changes if they don't save.
>       if (!FactoryUtils.savedBlockChanges(self)) {
> ```

This snippet of code is repeated a lot. You could add a util that does the alert and returns whether the user confirmed or cancelled.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/603#-:-KQ7jGlotEEzmZtnAfJw:bozd2x)*
<!-- Sent from Reviewable.io -->

vicng commented:
This looks really nice! It's a big improvement over what was there before.

---

Review status: 0 of 7 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[demos/blocklyfactory/app_controller.js, line 290 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8I2U2Gtfmv8V9Qeq7:-KQ8I2U2Gtfmv8V9Qeq8:b5cbw6r) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/app_controller.js#L290)):*
> ```JavaScript
>       this.selectedTab != AppController.BLOCK_FACTORY) {
>     // TODO: move savedBlockChanges to a Controller for the Block Factory Tab
>     if (!BlockFactory.isStarterBlock() &&
> ```

Rewrite into one if statement.

eg.

var hasUnsavedChanges = ...
if (hasUnsavedChanges &&
   !confirm(...)) {
}

---

*[demos/blocklyfactory/app_controller.js, line 530 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8IUgHYVFB4PE8szLe:-KQ8IUgHYVFB4PE8szLf:bgb1nkj) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/app_controller.js#L530)):*
> ```JavaScript
>       // warn user that proceeding to create a new block will cause them to lose
>       // their changes if they don't save.
>       if (!FactoryUtils.savedBlockChanges(self.blockLibraryController)) {
> ```

Rewrite as one if statement.

---

*[demos/blocklyfactory/block_library_controller.js, line 153 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8KrYpGxxu18WqudqK:-KQ8KrYqtQnfhBSjP8mz:buzj3j6) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L153)):*
> ```JavaScript
>   xmlElement.appendChild(Blockly.Xml.blockToDomWithXY(block));
> 
>   // Do not add agin option if block type is already in library.
> ```

Typo agin

---

*[demos/blocklyfactory/block_library_view.js, line 87 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8bUzQwsKWGxfMar5K:-KQ8bUzQwsKWGxfMar5L:b-xdlv40) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L87)):*
> ```JavaScript
>  */
> BlockLibraryView.prototype.setSelectedBlockType = function(blockTypeToSelect) {
>   if (this.optionMap[blockTypeToSelect]) {
> ```

Can the if/else logic be combined into one?

---

*[demos/blocklyfactory/block_library_view.js, line 118 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8bfZx14QgEacQzM7f:-KQ8bfZx14QgEacQzM7g:b-1y220h) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L118)):*
> ```JavaScript
>  *    a particular block type.
>  */
> BlockLibraryView.prototype.select_ = function(option) {
> ```

selectOption_ is better

---

*[demos/blocklyfactory/block_library_view.js, line 128 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8bo4acwZOTlVH3Y_K:-KQ8bo4acwZOTlVH3Y_L:bjqkepr) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L128)):*
> ```JavaScript
>  *    a particular block type.
>  */
> BlockLibraryView.prototype.deselect_ = function(option) {
> ```

And deselectOption_

---

*[demos/blocklyfactory/block_library_view.js, line 137 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8c8qjnpIEMuoS7duS:-KQ8c8qjnpIEMuoS7duT:b-2fq003) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L137)):*
> ```JavaScript
>  * whether the block is being saved or updated.
>  *
>  * @param {!Element} option - HTML 'a' element in the dropdown that represents
> ```

blockType?

---

*[demos/blocklyfactory/block_library_view.js, line 139 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8clp8v_aWDbp9BFwF:-KQ8clp8v_aWDbp9BFwG:b-1sdt05) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L139)):*
> ```JavaScript
>  * @param {!Element} option - HTML 'a' element in the dropdown that represents
>  *    a particular block type.
>  * @param {boolean} inLib - Whether the block type is in the library.
> ```

Abbreviations are evil :)

Change to "isInLibrary"?

---

*[demos/blocklyfactory/block_library_view.js, line 140 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8cnL_KGQAp-BS8wTo:-KQ8cnL_KGQAp-BS8wTp:b-4bv7mq) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L140)):*
> ```JavaScript
>  *    a particular block type.
>  * @param {boolean} inLib - Whether the block type is in the library.
>  * @param {boolean} saved - Whether changes to block have been saved.
> ```

savedChanges?

---

*[demos/blocklyfactory/block_library_view.js, line 184 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dEBUDodGZHqdPL4-:-KQ8dEBUDodGZHqdPL40:b7lh7im) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L184)):*
> ```JavaScript
>   } else {
>     // User has created new block.
>     this.saveButton.textContent = 'Save "block_type"';
> ```

Why "block_type"?

---

*[demos/blocklyfactory/factory.css, line 305 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dSUIPI3OjVBy2u06:-KQ8dSUIPI3OjVBy2u07:b9omrj6) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.css#L305)):*
> ```CSS
> 
> .button_alert {
>   background-color: #ffcccc;
> ```

Nit: Use 3 letter codes.

---

*[demos/blocklyfactory/factory.css, line 519 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dW_H4B3PI6C2S8eO:-KQ8dW_H4B3PI6C2S8eP:b-esr672) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.css#L519)):*
> ```CSS
> /* Change color of dropdown links on hover. */
> .dropdown-content a:hover, .dropdown-content label:hover {
>   background-color: #f1f1f1
> ```

Nit: 3 letter code/semicolon

---

*[demos/blocklyfactory/factory.css, line 524 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dXvLasKZqvdIQKpf:-KQ8dXvLasKZqvdIQKpg:b-esr672) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.css#L524)):*
> ```CSS
> /* Change color of dropdown links on selected. */
> .dropdown-content-selected {
>   background-color: #d8d8d8
> ```

Nit: 3 letter code/semicolon

---

*[demos/blocklyfactory/factory.js, line 214 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dgZMlyrd7p6ZOgqy:-KQ8dgZMlyrd7p6ZOgqz:b-6j12p1) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.js#L214)):*
> ```JavaScript
>       rootBlock.setWarningText('A standard Blockly.Block already exists ' +
>           'under this name.');
>     // Warn user to let them know they can't save a block under the default name
> ```

Move this comment inside the "else if" block.

---

*[demos/blocklyfactory/factory_utils.js, line 901 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8ecvn07Hd96XcmKtB:-KQ8ecvn07Hd96XcmKtC:b-lhe6v5) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory_utils.js#L901)):*
> ```JavaScript
> FactoryUtils.sameBlockXml = function(blockXml1, blockXml2) {
>   // Each block xml has only one child.
>   var blockXmlText1 = Blockly.Xml.domToText(
> ```

Will these potentially throw errors?

---

*[demos/blocklyfactory/index.html, line 294 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8eHFHdyh9bThnk4MD:-KQ8eHFHdyh9bThnk4ME:bed2wr) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/index.html#L294)):*
> ```HTML
>               Delete "block_type"
>             </button>
> <!--             <button id="createNewBlockButton" title="Create a new block.">
> ```

You can delete this now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/603#-:-KQ8fbvoxA-aIoAP_NLm:b-wxd0fy)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 7 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[demos/blocklyfactory/app_controller.js, line 290 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8I2U2Gtfmv8V9Qeq7:-KQH9pGq2lOUXjPx3SbI:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/app_controller.js#L290)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Rewrite into one if statement.
> 
> eg.
> 
> var hasUnsavedChanges = ...
> if (hasUnsavedChanges &&
>    !confirm(...)) {
> }
> 
> 
</details>
Done.

---

*[demos/blocklyfactory/app_controller.js, line 296 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7ek2M5-SOOcITbbET:-KQ8ZNMdPXmjB--AVehb:bz0q2qh) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/app_controller.js#L296)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> Do you mean = here?
</details>
Yes, thanks for catching that!

---

*[demos/blocklyfactory/app_controller.js, line 530 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8IUgHYVFB4PE8szLe:-KQHB555IhzTvKejKC4Q:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/app_controller.js#L530)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Rewrite as one if statement.
</details>
Done.

---

*[demos/blocklyfactory/block_library_controller.js, line 55 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7h3GfU95vQhbRBeaP:-KQ8ZukWIgyLH_VQvvWe:bpucmmy) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L55)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> I'm wondering if you need this. Mostly because I think the places where you use it (e.g. hiding/showing the list), you probably should just have a view method for hide/show.  
</details>
Removed and moved the reference to the dropdownID to the view

---

*[demos/blocklyfactory/block_library_controller.js, line 153 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8KrYpGxxu18WqudqK:-KQHBHaK2qZ8ceD3y1lQ:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L153)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Typo agin
</details>
Done.

---

*[demos/blocklyfactory/block_library_controller.js, line 190 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7hbh7CSJnxPJcHS4S:-KQ8axfzB-C63QWw-Ocu:bksoywh) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L190)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> should this be done by the view? (caveat: I haven't looked at what addoptionselectHandlers() does yet)
</details>
I keep this here because the click handlers deal with not only the block library view (the dropdown) but also the block factory view (the main workspace). on select of a block lib dropdown option, the we access the block library storage in order to get the appropriate xml that gets opened on the blockfactory main workspace

---

*[demos/blocklyfactory/block_library_controller.js, line 285 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7g4UvW_TxjYrbqG2U:-KQHY_NiHI25fo_Y7-HM:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L285)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> This snippet of code is repeated a lot. You could add a util that does the alert and returns whether the user confirmed or cancelled.
</details>
Done.

---

*[demos/blocklyfactory/block_library_view.js, line 87 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8bUzQwsKWGxfMar5K:-KQH_J9CzilWKcsjebgG:bvibv9a) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L87)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Can the if/else logic be combined into one?
</details>
super simplified this ! thanks for bringing this up. The updating of the buttons is done with a eventlistener on the BlockFactory.mainWorkspace so the buttons don't need to be updated here. as a result, i've removed that code. removing that code meant that there wasn't anything to distinguish the behaviors based on the different cases. woo!

---

*[demos/blocklyfactory/block_library_view.js, line 118 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8bfZx14QgEacQzM7f:-KQH_dlD4Dukn3_WRKWs:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L118)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> selectOption_ is better
</details>
Done.

---

*[demos/blocklyfactory/block_library_view.js, line 128 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8bo4acwZOTlVH3Y_K:-KQHBaqZPWAiTe0beWRP:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L128)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> And deselectOption_
</details>
Done.

---

*[demos/blocklyfactory/block_library_view.js, line 137 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8c8qjnpIEMuoS7duS:-KQHBoJx2m0P8suo3C4B:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L137)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> blockType?
</details>
Done.

---

*[demos/blocklyfactory/block_library_view.js, line 139 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8clp8v_aWDbp9BFwF:-KQHBrtdupdw6_YfApu1:b-43wz42) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L139)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Abbreviations are evil :)
> 
> Change to "isInLibrary"?
</details>
LOL okay! i thought the abbreviations are evil only applied to the function names. now I know. fixed!!

---

*[demos/blocklyfactory/block_library_view.js, line 140 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8cnL_KGQAp-BS8wTo:-KQHC6e2h8kQD9unZmDs:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L140)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> savedChanges?
</details>
Done.

---

*[demos/blocklyfactory/block_library_view.js, line 184 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dEBUDodGZHqdPL4-:-KQHCCz6VbcDDVUf2n5B:b-xirjoh) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L184)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Why "block_type"?
</details>
I actually deleted this code because  in all uses of updateButtons, a blockType is passed to the function. Thus this code never executes. The handling of 'block_type' is in the above if/else statement. 

Thanks for raising this question!

---

*[demos/blocklyfactory/factory.css, line 305 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dSUIPI3OjVBy2u06:-KQHI9mR5wcy_cxdbh45:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.css#L305)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Nit: Use 3 letter codes.
</details>
Done.

---

*[demos/blocklyfactory/factory.css, line 519 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dW_H4B3PI6C2S8eO:-KQHIo4ayomttwBPl-lB:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.css#L519)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Nit: 3 letter code/semicolon
</details>
Done.

---

*[demos/blocklyfactory/factory.css, line 524 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dXvLasKZqvdIQKpf:-KQHJYOCAx1rjnmMQN1w:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.css#L524)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Nit: 3 letter code/semicolon
</details>
Done.

---

*[demos/blocklyfactory/factory.js, line 214 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dgZMlyrd7p6ZOgqy:-KQHa7KK9K2PjNcgUR1B:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.js#L214)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Move this comment inside the "else if" block.
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 901 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8ecvn07Hd96XcmKtB:-KQHRbAa-jRB3QN5gkr0:b2w1p6r) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory_utils.js#L901)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Will these potentially throw errors?
</details>
Yes, I've split up the code and checked for 1) that the things provided ARE xml and 2) that they contain a block element. 

---

*[demos/blocklyfactory/index.html, line 294 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8eHFHdyh9bThnk4MD:-KQHK53A0j4W37CBEOzg:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/index.html#L294)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> You can delete this now.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/603)*
<!-- Sent from Reviewable.io -->

evd2014 commented:
Review status: 0 of 7 files reviewed at latest revision, 22 unresolved discussions.

---

*[demos/blocklyfactory/factory_utils.js, line 917 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQMHluoeGEAwdZn2q8r:-KQMHlupZyx_mEXnN1z3:b-ze2ve6) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory_utils.js#L917)):*
> ```JavaScript
> /**
>  * Returns whether or not a block's changes has been saved to the Block Library.
>  * TODO: move into the Block Factory Controller once made.
> ```

Assign TODO

---

*[demos/blocklyfactory/factory_utils.js, line 919 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQMHn4wnQOoTzmkBEQx:-KQMHn4wnQOoTzmkBEQy:b-iuzbqi) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory_utils.js#L919)):*
> ```JavaScript
>  * TODO: move into the Block Factory Controller once made.
>  *
>  * @param {!BlockLibraryController} blockLibraryController - Block Library Controller storing
> ```

nit: > 80 characters

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/603)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 7 files reviewed at latest revision, 22 unresolved discussions.

---

*[demos/blocklyfactory/factory_utils.js, line 917 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQMHluoeGEAwdZn2q8r:-KQMKistjD0vSL8qO0sC:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory_utils.js#L917)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Assign TODO
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 919 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQMHn4wnQOoTzmkBEQx:-KQMKj7xSaKjrb3cPoxl:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory_utils.js#L919)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> nit: > 80 characters
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/603)*
<!-- Sent from Reviewable.io -->

vicng commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, 23 unresolved discussions.

---

*[demos/blocklyfactory/app_controller.js, line 530 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8IUgHYVFB4PE8szLe:-KQNCBeEF7KKutyPRsQt:b-78adxt) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/app_controller.js#L530)):*
<details><summary><i>Previously, quachtina96 wrote…</i></summary>
> Done.
</details>
You can just use FactoryUtils.warnIfUnsavedChanges() now.

---

*[demos/blocklyfactory/block_library_controller.js, line 285 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7g4UvW_TxjYrbqG2U:-KQNulJIBusSqk6hnnn7:b-739vhf) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L285)):*
<details><summary><i>Previously, quachtina96 wrote…</i></summary>
> Done.
</details>
This can be part of the if statement now too.

---

*[demos/blocklyfactory/block_library_view.js, line 87 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8bUzQwsKWGxfMar5K:-KQNvFjgGq2tcZW-1bY-:b-83sjoh) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_view.js#L87)):*
<details><summary><i>Previously, quachtina96 wrote…</i></summary>
> super simplified this ! thanks for bringing this up. The updating of the buttons is done with a eventlistener on the BlockFactory.mainWorkspace so the buttons don't need to be updated here. as a result, i've removed that code. removing that code meant that there wasn't anything to distinguish the behaviors based on the different cases. woo!
</details>
Nice!

---

*[demos/blocklyfactory/factory_utils.js, line 901 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8ecvn07Hd96XcmKtB:-KQNvl5_FFhiyH8eOImr:beqesph) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory_utils.js#L901)):*
<details><summary><i>Previously, quachtina96 wrote…</i></summary>
> Yes, I've split up the code and checked for 1) that the things provided ARE xml and 2) that they contain a block element. 
</details>
You don't necessarily need to throw an error from this method (it could just return false). Totally up to you what the behaviour is tho, I'm fine either way.

---

*[demos/blocklyfactory/index.html, line 283 \[r6\]](https://reviewable.io:443/reviews/google/blockly/603#-KQNw1uKnNv2kPphfYIN:-KQNw1uKnNv2kPphfYIO:bhlrlxb) ([raw file](https://github.com/google/blockly/blob/7fdbbf74ed5568d4c9d9cf795a37b174e504f79d/demos/blocklyfactory/index.html#L283)):*
> ```HTML
>             <td id="blockLibraryContainer">
>             <span>
> <!--               <h3>Block Library:</h3> -->
> ```

Delete comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/603#-:-KQNw6oYfAs_HikYeJ1T:bnfp4nl)*
<!-- Sent from Reviewable.io -->

picklesrus commented:
Review status: 0 of 7 files reviewed at latest revision, 25 unresolved discussions.

---

*[demos/blocklyfactory/block_library_controller.js, line 269 \[r6\]](https://reviewable.io:443/reviews/google/blockly/603#-KQNwtntYWXxr_zX0D5r:-KQNwtnu-vuAqJYT-ENF:b3tewet) ([raw file](https://github.com/google/blockly/blob/7fdbbf74ed5568d4c9d9cf795a37b174e504f79d/demos/blocklyfactory/block_library_controller.js#L269)):*
> ```JavaScript
>     self.view.setSelectedBlockType(blockType);
>     self.openBlock(blockType);
>     //The block is saved in the block library and all changes have been saved
> ```

space between // and The. Same on the last 2 lines of the comment.

---

*[demos/blocklyfactory/block_library_controller.js, line 270 \[r6\]](https://reviewable.io:443/reviews/google/blockly/603#-KQNx0OBZsfNRQ0egh84:-KQNx0OBZsfNRQ0egh85:b-dg5x7) ([raw file](https://github.com/google/blockly/blob/7fdbbf74ed5568d4c9d9cf795a37b174e504f79d/demos/blocklyfactory/block_library_controller.js#L270)):*
> ```JavaScript
>     self.openBlock(blockType);
>     //The block is saved in the block library and all changes have been saved
>     // when the users opens a block from the block library dropdown.
> ```

s/users/user. 

---

*[demos/blocklyfactory/factory.js, line 214 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dgZMlyrd7p6ZOgqy:-KQNyTJbGpxyJw60JBwo:bckun3z) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.js#L214)):*
<details><summary><i>Previously, quachtina96 (Tina Quach) wrote…</i></summary>
> Done.
</details>
nit: now it is > 80

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/603)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 7 files reviewed at latest revision, 25 unresolved discussions.

---

*[demos/blocklyfactory/block_library_controller.js, line 285 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ7g4UvW_TxjYrbqG2U:-KQ8cWjwbxfPZsdIeFmO:b-goltaq) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/block_library_controller.js#L285)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> This can be part of the if statement now too.
</details>
I prefer it as is because it provides more clarity on what warnIfUnsavedChanges is returning

---

*[demos/blocklyfactory/block_library_controller.js, line 269 \[r6\]](https://reviewable.io:443/reviews/google/blockly/603#-KQNwtntYWXxr_zX0D5r:-KQOq00LE7sok4QnhQmI:b-896fix) ([raw file](https://github.com/google/blockly/blob/7fdbbf74ed5568d4c9d9cf795a37b174e504f79d/demos/blocklyfactory/block_library_controller.js#L269)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> space between // and The. Same on the last 2 lines of the comment.
</details>
Done.

---

*[demos/blocklyfactory/block_library_controller.js, line 270 \[r6\]](https://reviewable.io:443/reviews/google/blockly/603#-KQNx0OBZsfNRQ0egh84:-KQOqAvgJsUAIRaBFAKi:b-896fix) ([raw file](https://github.com/google/blockly/blob/7fdbbf74ed5568d4c9d9cf795a37b174e504f79d/demos/blocklyfactory/block_library_controller.js#L270)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> s/users/user. 
</details>
Done.

---

*[demos/blocklyfactory/factory.js, line 214 \[r3\]](https://reviewable.io:443/reviews/google/blockly/603#-KQ8dgZMlyrd7p6ZOgqy:-KQOqbnZuWgnHXNbjF9b:b-896fix) ([raw file](https://github.com/google/blockly/blob/665e23aa6b1b65a34f4a5d55a6cd4ca27c9f101c/demos/blocklyfactory/factory.js#L214)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary>
> nit: now it is > 80
</details>
Done.

---

*[demos/blocklyfactory/index.html, line 283 \[r6\]](https://reviewable.io:443/reviews/google/blockly/603#-KQNw1uKnNv2kPphfYIN:-KQOqq19t267-kyfqJN3:b-896fix) ([raw file](https://github.com/google/blockly/blob/7fdbbf74ed5568d4c9d9cf795a37b174e504f79d/demos/blocklyfactory/index.html#L283)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Delete comment.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/603)*
<!-- Sent from Reviewable.io -->

picklesrus commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, 25 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/603#-:-KQREgQ1syL3m8m68Voi:bnfp4nl)*
<!-- Sent from Reviewable.io -->
